File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
def solution(new_id):
    answer = ''
    dot_flag=False
        
    for idx, ch in enumerate(new_id):        
        if ch.isupper():
            answer+=ch.lower()
            continue
            
        if ch.islower() or ch.isdigit() or ch in ('-','_','.'):
            answer+=ch
            
    while '..' in answer:
        answer = answer.replace('..', '.')
    
    answer = answer.lstrip('.')
    answer = answer.rstrip('.')
            
    if len(answer)==0:
        answer = 'a'
    
    if len(answer)>=16:
        answer = answer[:15].rstrip('.')
        
    if len(answer)<=2:
        while True:
            last = answer[-1]
            answer+=last
            if len(answer)==3: break
    
    return answer

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.java
import java.util.*;

class Solution {
    public String solution(String new_id) {
        StringBuilder answer = new StringBuilder();
        
        for (int i=0; i<new_id.length(); i++) {
            char ch = new_id.charAt(i);
            if (Character.isUpperCase(ch)) {
                answer.append(Character.toLowerCase(ch));
            }
            else if (Character.isLowerCase(ch)||Character.isDigit(ch)||ch=='-'||ch=='_'||ch=='.') {
                answer.append(ch);
            }                  
        }
        
        String result = answer.toString();
        
        while (result.contains("..")) {
            result = result.replace("..",".");
        }
        
        if (result.startsWith(".")) result = result.substring(1);
        if (result.endsWith(".")) result = result.substring(0, result.length()-1);
        if (result.length()==0) result += 'a';
        if (result.length()>=16) result = result.substring(0, 15);
        if (result.endsWith(".")) result = result.substring(0, result.length()-1);
        if (result.length()<=2) {
            while (true) {
                result+=result.substring(result.length()-1);
                if (result.length()==3) break;
            }
        }
        
        return result;
    }
}